### PR TITLE
Add more generic EdDSA tests

### LINKED ISSUE #272: Test coverage for key comparison and getting public key data from EdDSA
The PR #267 added key comparison test `tcmpkeys` which runs for RSA and EC keys, but there is no test coverage for EdDSA keys. The implementation should be complete though.

It also added reading of public attributes from the generated keys in `tgenkey`, but also without EdDSA coverage, which needs to be added somehow separately (as the EdDSA is available only in SoftHSM for now).

BTW when is the NSS getting EdDSA support? There were some patches, werent they?

### PR DESCRIPTION
**Pull request by simo5**
#### Description

Implements tests requested in #272

Fixes #272

#### Checklist

- [x] Code modified for feature
- [x] Test suite updated with functionality tests
- [ ] ~Test suite updated with negative tests~
- [ ] ~Documentation updated~

#### Reviewer's checklist:

- [x] Any issues marked for closing are addressed
- [x] There is a test suite reasonably covering new functionality or modifications
- [x] This feature/change has adequate documentation added
- [x] Code conform to coding style that today cannot yet be enforced via the check style test
- [x] Commits have short titles and sensible commit messages
- [ ] Coverity Scan has run if needed (code PR) and no new defects were found

### CODE REVIEW

**Comment by simo5** [comment]
@Jakuje I stole (and rebased) one of your commits from #292 but otherwise just implemented basic test via changes to tcmpkeys and tgenkeys so they could be used in tests/tedwards

My last commit changes tgenkey to remove custom detection given my general tgenkey changes allow us to call ED25519 checks only from the appropriate test script.

**Comment by simo5** [comment]
Lovely, address sanitizer fails w/o a peep in the logs ... will look at it next week

File: tests/tcmpkeys.c [hunk 1]
```diff
@@ -44,6 +44,14 @@ static EVP_PKEY *load_key(const char *uri)
         exit(EXIT_FAILURE);
     }
 
+    if (strncmp(uri, "pkcs11:", 7)) {
+        /* public file */
+        if (OSSL_STORE_expect(store, OSSL_STORE_INFO_PUBKEY) != 1) {
+            fprintf(stderr, "Failed to expect Public Key File\n");
+            exit(EXIT_FAILURE);
+        }
+    }
```
Comment (Jakuje) [inline tests/tcmpkeys.c:53]: I do not think this is correct. The uris in the test below are `EDPRIURI` or `EDPUBURI`, so we should be interested in getting the private key when requested (which is broken in openssl 3.1). I think this is just a workaround for the OpenSSL issue openssl/openssl#22246 (and openssl/openssl#21219) as in most of the useful cases, we are interested in the private key to be able to use it for something sensible.

This significantly limits the testing. I do not mind adding another option in the tcmpkeys to check we get public key, but we should be able to get a private key too.

File: tests/tgenkey.c [hunk 2]
```diff
@@ -313,139 +345,220 @@ static void sign_test(const char *label, bool fail)
...
-    ret = asprintf(&uri, "pkcs11:object=%s;id=%s", label, idhex);
-    if (ret == -1) {
-        fprintf(stderr, "Failed to make label");
-        exit(EXIT_FAILURE);
+            free(label);
+            free(uri);
+        } else if (strcmp(tests[num], "ED25519") == 0) {
+            ret = RAND_bytes(id, 16);
+            if (ret != 1) {
+                fprintf(stderr, "Failed to set generate key id\n");
```
Comment (Jakuje) [inline tests/tgenkey.c:533]:
```suggestion
                fprintf(stderr, "Failed to generate key id\n");
```

File: tests/tgenkey.c [hunk 3]
```diff
@@ -313,139 +345,220 @@ static void sign_test(const char *label, bool fail)
...
+            miniid = (id[0] << 24) + (id[1] << 16) + (id[2] << 8) + id[3];
+            ret = asprintf(&label, "Test-Ed-gen-%08x", miniid);
+            if (ret == -1) {
+                fprintf(stderr, "Failed to make label\n");
+                exit(EXIT_FAILURE);
+            }
+            hexify(idhex, id, 16);
+            ret = asprintf(&uri, "pkcs11:object=%s;id=%s", label, idhex);
+            if (ret == -1) {
+                fprintf(stderr, "Failed to make label\n");
```
Comment (Jakuje) [inline tests/tgenkey.c:545]:
```suggestion
                fprintf(stderr, "Failed to make PKCS #11 URI\n");
```

Comment (Jakuje) [reply tests/tgenkey.c:545]: The same in the other key types.

File: tests/tcmpkeys.c (hunk 1 above)
Comment (simo5) [reply tests/tcmpkeys.c]: I pushed a commit that should address these concerns.

**Review by Jakuje — APPROVED** [review]

**Comment by simo5** [comment]
Thank you!

**Pull request merged**